lib/internal/modules/esm/resolve.js

```diff
@@ -153,44 +160,29 @@ function fileExists(url) {
  * @returns {URL}
  */
 function legacyMainResolve(packageJSONUrl, packageConfig, base) {
-  let guess;
-  if (packageConfig.main !== undefined) {
-    // Note: fs check redundances will be handled by Descriptor cache here.
-    if (fileExists(guess = new URL(`./${packageConfig.main}`,
-                                   packageJSONUrl))) {
-      return guess;
-    } else if (fileExists(guess = new URL(`./${packageConfig.main}.js`,
-                                          packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}.json`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}.node`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.js`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.json`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.node`,
-                                        packageJSONUrl)));
-    else guess = undefined;
-    if (guess) {
-      emitLegacyIndexDeprecation(guess, packageJSONUrl, base,
-                                 packageConfig.main);
-      return guess;
+  const resolvedOption = URLLegacyMainResolve(packageJSONUrl.href, packageConfig.main);
+
+  // less than zero means that happened some error
+  // -2: LegacyMainResolveReturnType::kInvalidUrl
+  // -1: LegacyMainResolveReturnType::kModuleNotFound
+  if (resolvedOption < 0) {
+    if (resolvedOption === -2) {
+      throw new ERR_INVALID_URL(packageConfig.main || './index.js');
+    } else {
+      throw new ERR_MODULE_NOT_FOUND(
+        fileURLToPath(new URL('.', packageJSONUrl)), fileURLToPath(base));
     }
-    // Fallthrough.
   }
-  if (fileExists(guess = new URL('./index.js', packageJSONUrl)));
-  // So fs.
-  else if (fileExists(guess = new URL('./index.json', packageJSONUrl)));
-  else if (fileExists(guess = new URL('./index.node', packageJSONUrl)));
-  else guess = undefined;
-  if (guess) {
-    emitLegacyIndexDeprecation(guess, packageJSONUrl, base, packageConfig.main);
-    return guess;
-  }
-  // Not found.
-  throw new ERR_MODULE_NOT_FOUND(
-    fileURLToPath(new URL('.', packageJSONUrl)), fileURLToPath(base));
+
+  // 0-6: when packageConfig.main is defined
+  // 7-9: when packageConfig.main is NOT defined,
+  //      or when the previous case didn't found the file
+  const baseUrl = resolvedOption <= 6 ? `./${packageConfig.main}` : '';
```

Comment:
I'd compare it with a variable instead, just to make it more readable. Something like:

`isPackageConfigMainDefined(resolvedOption)`